### PR TITLE
Integrate Engram into custom model

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

Integrate Engram feature into a custom model
* Add configs into `base.yml` and `type.py`, integrates with `deepseek-custom` model
* This PR supports unscan version of Engram, and scan version will be next PR.
* Tried to initialize the hash map to model level for one time initialization, but met various JAX initialization error (see more in b/478294699 and this [PR](https://github.com/AI-Hypercomputer/maxtext/pull/3125)). Also added a comment there.
* Currently, to make it work, you will see mix of `jnp` and `np` as some operations running on CPU to avoid ConcretizationTypeError and other issues. 

# Tests

* Expect github runners to pass
* Unit tests still passing for Engram: [link](https://paste.googleplex.com/4533027603611648)
* End-to-end unscan training for custom model: [link](https://paste.googleplex.com/6288392850636800)
* Sanity check for DS v2 (expect no impact)

```
Before change:

I0218 19:13:51.139775 139970093350464 max_utils.py:697] 	Using (GB) 43.98 / 95.74 (45.936912%) on TPU_1(process=0,(1,0,0,0))
I0218 19:15:18.498541 139970093350464 metric_logger.py:181] completed step: 19, seconds: 4.366, TFLOP/s/device: 123.145, Tokens/s/device: 7505.830, total_weights: 131072, loss: 8.135

After change:

I0218 19:05:35.490278 140385131265600 max_utils.py:697] 	Using (GB) 43.99 / 95.74 (45.947357%) on TPU_0(process=0,(0,0,0,0))
I0218 19:07:02.849351 140385131265600 metric_logger.py:181] completed step: 19, seconds: 4.366, TFLOP/s/device: 123.144, Tokens/s/device: 7505.806, total_weights: 131072, loss: 8.135
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/models/deepseek.py [hunk 1]
```diff
@@ -91,6 +97,43 @@ def __init__(
...
+          rngs=rngs,
+      )
+      tokenizer = transformers.AutoTokenizer.from_pretrained(config.tokenizer_path, token=config.hf_access_token)
+      # TODO(ranran): Refactor NgramHashMapping to initialize once globally or at the model level.
+      # Moving this to decoders.py currently causes JAX initialization errors.
+      self.ngram_hash_mapping = NgramHashMapping(
+          engram_vocab_bases=config.engram_vocab_sizes,
+          max_ngram_size=config.engram_max_ngram_size,
+          engram_num_heads=config.engram_num_heads,
+          layer_ids=[layer_idx],
```
Comment (shuningjin) [inline src/maxtext/models/deepseek.py]: This should be initialized with global layer_ids, as in unit test. Otherwise hashing behavior is different.
https://github.com/AI-Hypercomputer/maxtext/blob/52e0ceafa5f65f8e99c44cf19acca90439f33d3a/tests/unit/engram_vs_reference_test.py#L762

File: src/maxtext/configs/base.yml [hunk 2]
```diff
@@ -1093,3 +1094,20 @@ force_q_layout: false
...
+################################## DeepSeek Engram ##################################
+# Indices of transformer layers where Engram are integrated; leave empty [] to disable.
+# Example: [1, 4] attaches to the 2nd and 5th layer. 
+engram_layers: []
+# Number of heads dedicated to the Engram.
+engram_num_heads: 8
+# Head dimension for heads.
+engram_head_dim: 1280
+# List of vocab sizes for each n-gram order.
+engram_vocab_sizes: []
```
Comment (shuningjin) [inline src/maxtext/configs/base.yml]: (1) This config can be confusing: This is not the actual size, but the search start for prime sizes.  

For accuracy, I suggest: `engram_vocab_bases`, and `# List of minimum head vocab sizes for each n-gram order in 2...engram_max_ngram_size`.

For example, if the `max_ngram = 2, num_head=3, engram_vocab_bases = [4]`
- ngram2, head0, find_next_unseen_prime > 4-1  -- size = 5
- ngram2, head1, find_next_unseen_prime > 5 --- size = 7
- ngram2, head2, find_next_unseen_prime > 7 --- size = 11

(2) to reflect the hierarchy (engram -> 2...max ngram -> num_head per ngram), the order could be:
```
engram_max_ngram_size 
engram_num_heads, engram_head_dim
engram_vocab_sizes
```

File: src/maxtext/configs/types.py [hunk 3]
```diff
@@ -2247,6 +2264,13 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
         and self.gradient_accumulation_steps > 1
     ):
       raise ValueError("FP8 quantization is not compatible with gradient accumulation.")
+    if self.engram_layers:
```
Comment (shuningjin) [inline src/maxtext/configs/types.py:2284]: Thanks for adding the checks! Might be good to check `len(engram_vocab_sizes) == (engram_max_ngram_size -1)`

File: src/MaxText/layers/engram.py [hunk 4]
```diff
@@ -181,10 +182,14 @@ def __init__(
     self.compressed_tokenizer = CompressedTokenizer(tokenizer)
     self.tokenizer_vocab_size = len(self.compressed_tokenizer)
     if pad_id is not None:
```
Comment (shuningjin) [inline src/MaxText/layers/engram.py]: When pad_id is not none, we init self.pad_id, which is required by _get_ngram_hashes for padding.

When it is None, it is unclear from reference
https://github.com/deepseek-ai/Engram/blob/fb7f84a21f91223715394a33a1dc24bbfb7f788e/engram_demo_v1.py#L211
Maybe raise error, and add a todo in comment?

File: src/maxtext/models/deepseek.py [hunk 5]
```diff
@@ -91,6 +97,43 @@ def __init__(
...
+          num_features=self.dummy_inputs_shape[-1],
+          dtype=self.config.dtype,
+          weight_dtype=self.config.weight_dtype,
+          kernel_axes=("norm",),
+          epsilon=self.config.normalization_layer_epsilon,
+          rngs=rngs,
+      )
+      tokenizer = transformers.AutoTokenizer.from_pretrained(config.tokenizer_path, token=config.hf_access_token)
+      # TODO(ranran): Refactor NgramHashMapping to initialize once globally or at the model level.
+      # Moving this to decoders.py currently causes JAX initialization errors.
```
Comment (shuningjin) [inline src/maxtext/models/deepseek.py:112]: Alternatively, can we initialize it in data pipeline? treating it as a special tokenizer, operating with np and cpu, similar to existing hf_tokenizer

Comment (RissyRan) [reply src/maxtext/models/deepseek.py:112]: Yeah, I tried both in this draft [PR](https://github.com/AI-Hypercomputer/maxtext/pull/3125):

1) put in the data pipeline: https://screenshot.googleplex.com/5dhowKp7EjzMuud - not working.
2) put in the model level: https://screenshot.googleplex.com/5FW7nxAXu9fFDCA - not working.

Met different issues here and there. I think if bandwidth allows, we could dive deep more.

File: src/maxtext/models/deepseek.py (hunk 1 above)
Comment (RissyRan) [reply src/maxtext/models/deepseek.py]: Yes, this is global version: https://screenshot.googleplex.com/4pEmyDgFWhiXF6c

File: src/MaxText/layers/engram.py (hunk 4 above)
Comment (RissyRan) [reply src/MaxText/layers/engram.py]: OK, let me raise am error here.

File: src/maxtext/configs/base.yml (hunk 2 above)
Comment (RissyRan) [reply src/maxtext/configs/base.yml]: Yeah, it seems this is the start value for the current prime: https://github.com/deepseek-ai/Engram/blob/fb7f84a21f91223715394a33a1dc24bbfb7f788e/engram_demo_v1.py#L246.

Let me update a little bit.

**Comment by RissyRan** [comment]
> Thank you! I have some minor comment. As a follow up, we could try placing ngram-hash-mapping globally for efficiency: either (1) at model level or (2) at data pipeline level, treat it as a special tokenizer.

Thanks for the review! As noted in the PR description and our previous [discussion](https://screenshot.googleplex.com/6zPforq3i3npFrq), I’d like to prioritize getting the core functionality landed first. We can then tackle those trickier edge cases in a follow-up. How does that sound?

File: src/maxtext/models/deepseek.py (hunk 1 above)
Comment (shuningjin) [reply src/maxtext/models/deepseek.py]: To clarify, if `self.config.engram_layer_ids = [1, 15]`, both layer should have `layer_ids=[1, 15]`, rather than `layer_ids=[1]` and `layer_ids[15]` respectively. 

[Vocab size search](https://github.com/deepseek-ai/Engram/blob/fb7f84a21f91223715394a33a1dc24bbfb7f788e/engram_demo_v1.py#L236-L239) needs to know all layers at once. It affects the `seen_prime`: if layer1 takes prime 3, layer 15 cannot use that size. See this small example: https://paste.googleplex.com/6291527186907136.

Comment (RissyRan) [reply src/maxtext/models/deepseek.py]: Sounds good, updated.

File: tests/unit/engram_vs_reference_test.py
Comment (shuningjin) [inline tests/unit/engram_vs_reference_test.py:792]: Could you update these `self.config` to `cfg` in class initialization to align unit test with our e2e? 
- For example, `self.config.engram_layer_ids` -> `cfg.engram_layers`. 
- May need to change `Config` in the top to align names. 
- Same for other tests in this file.

similar to 
https://github.com/AI-Hypercomputer/maxtext/blob/d8c8862f9eff45296e7d358a5c9c7bb062ff35de/tests/unit/deepseek32_vs_reference_test.py#L972-L989

Comment (RissyRan) [reply tests/unit/engram_vs_reference_test.py:792]: Sure, I see mainly is `engram_layers` change. Updated and tested.

```
.
----------------------------------------------------------------------
Ran 8 tests in 51.458s

OK
```

File: src/maxtext/configs/types.py [hunk 6]
```diff
@@ -2250,6 +2269,18 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
...
+      if not self.hf_access_token or not self.tokenizer_path:
+        assert ValueError(
```
Comment (gagika) [inline src/maxtext/configs/types.py]: assert -> raise

File: src/maxtext/configs/types.py [hunk 7]
```diff
@@ -2250,6 +2269,18 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
...
+            "Engram requires both 'hf_access_token' and 'tokenizer_path' " "to load the Hugging Face tokenizer."
+        )
+      if self.scan_layers:
+        assert NotImplementedError("Currently Engram only supports unscanned version. Please set scan_layers=False.")
```
Comment (gagika) [inline src/maxtext/configs/types.py]: assert -> raise

File: src/maxtext/layers/engram.py [hunk 8]
```diff
@@ -118,19 +120,19 @@ def _build_lookup_table(self, tokenizer: HFTokenizer, normalizer: normalizers.Se
 
     return old2new, len(key2new)
 
-  def __call__(self, input_ids) -> np.ndarray:
+  def __call__(self, input_ids) -> Array:
     """
     Maps original token IDs to compressed IDs.
     """
-    input_ids = np.asarray(input_ids, dtype=np.int64)
+    input_ids = jnp.asarray(input_ids, dtype=jnp.int64)
```
Comment (gagika) [inline src/maxtext/layers/engram.py:127]: is jnp.int64 really needed? or `jnp.int32` is good enough for table sizes we will need?

File: src/maxtext/layers/engram.py [hunk 9]
```diff
@@ -118,19 +120,19 @@ def _build_lookup_table(self, tokenizer: HFTokenizer, normalizer: normalizers.Se
...
 
     # Identify valid tokens (ignore padding/masks usually marked with negative IDs)
-    valid_mask = input_ids >= 0
-    valid_ids = input_ids[valid_mask]
+    # In JAX, we map negative IDs to 0 (or a valid index) for lookup, then mask output back.
```
Comment (gagika) [inline src/maxtext/layers/engram.py]: should the comment be updated?

File: src/maxtext/configs/types.py (hunk 6 above)
Comment (RissyRan) [reply src/maxtext/configs/types.py]: Sounds good!

File: src/maxtext/layers/engram.py (hunk 8 above)
Comment (RissyRan) [reply src/maxtext/layers/engram.py:127]: Good point. My initial attempt with jnp.int32 failed due to a mismatch with PyTorch's int64, but we could try unifying both to int32. The range (2^31-1) is should be sufficient for our vocab and hash IDs. @shuningjin, any concerns?

File: src/maxtext/layers/engram.py (hunk 9 above)
Comment (RissyRan) [reply src/maxtext/layers/engram.py]: I think so? I also update the code a little bit bellow.

File: src/maxtext/layers/engram.py (hunk 8 above)
Comment (RissyRan) [reply src/maxtext/layers/engram.py:127]: Also discussing offline for test failures by converting int64 to int32.

File: src/maxtext/layers/engram.py (hunk 9 above)
Comment (RissyRan) [reply src/maxtext/layers/engram.py]: I updated a little bit. Could you explain more if haven't addressed it yet.

Comment (gagika) [reply src/maxtext/layers/engram.py]: looks good now, previous comment was refering to refactored code.
